src/main/java/org/cryptomator/cloudaccess/requestdecorator/QuotaRequestCachingDecorator.java

```diff
@@ -0,0 +1,51 @@
+package org.cryptomator.cloudaccess.requestdecorator;
+
+import com.github.benmanes.caffeine.cache.AsyncCache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import org.cryptomator.cloudaccess.api.CloudPath;
+import org.cryptomator.cloudaccess.api.CloudProvider;
+import org.cryptomator.cloudaccess.api.Quota;
+import org.cryptomator.cloudaccess.api.exceptions.NotFoundException;
+import org.cryptomator.cloudaccess.api.exceptions.QuotaNotAvailableException;
+
+import java.time.Duration;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * Decorates an existing CloudProvider by caching quota-requests for a duration of default 10 seconds (can be set using <code>org.cryptomator.cloudaccess.metadatacachingprovider.timeoutSeconds</code>).
+ */
+class QuotaRequestCachingDecorator implements CloudProviderDecorator {
+
+	private final static int DEFAULT_CACHE_TIMEOUT_SECONDS = 10;
+
+	// visible for testing
+	final AsyncCache<CloudPath, Quota> quotaCache;
+
+	private final CloudProvider delegate;
+
+	public QuotaRequestCachingDecorator(CloudProvider delegate) {
+		this(delegate, Caffeine
+				.newBuilder()
+				.expireAfterWrite(Duration.ofSeconds(Integer.getInteger("org.cryptomator.cloudaccess.metadatacachingprovider.timeoutSeconds", DEFAULT_CACHE_TIMEOUT_SECONDS)))
+				.buildAsync());
+	}
+
+	QuotaRequestCachingDecorator(CloudProvider delegate, AsyncCache<CloudPath, Quota> quotaCache) {
+		this.delegate = delegate;
+		this.quotaCache = quotaCache;
+	}
+
+	@Override
+	public CloudProvider delegate() {
+		return delegate;
+	}
+
+	@Override
+	public CompletionStage<Quota> quota(CloudPath folder) {
+		return quotaCache.get(folder, k -> delegate.quota(k).whenComplete((metadata, throwable) -> {
+			if (throwable != null && !(throwable instanceof NotFoundException) && !(throwable instanceof QuotaNotAvailableException)) {
+				quotaCache.synchronous().invalidate(folder);
+			}
+		}).toCompletableFuture().join());
```

Comment:
I suspect you meant to use `get(key, bifunction)` to return the future directly the cache and avoid the `.join()` / wrapper.

Note that if the future Caffeine gets throws an exception or returns null then it will automatically discard the entry in its `whenComplete` handler. Of course if you want to cache failures this is akin to "negative caching", e.g. an `Optional<V>` or `Try<V>` wrapper. Then the cache can have a successful value, but you can unwrap / transform /  return a failed future upstream to the caller.

Reply:
Thanks for the hint, just fixed it in https://github.com/cryptomator/cloud-access-java/commit/7242de32c503d00430bb473a8a1c12f38bddf427